src/modules/flow/adafruit/adafruit.c

```diff
@@ -0,0 +1,161 @@
+/*
+ * This file is part of the Soletta Project
+ *
+ * Copyright (C) 2015 Intel Corporation. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "adafruit-gen.h"
+#include "sol-flow-internal.h"
+
+#include <sol-spi.h>
+#include <sol-util.h>
+#include <errno.h>
+
+
+struct lcd_strip_data {
+    struct sol_spi *spi;
+    int last_set_pixel;
+    int last_set_color;
+    uint16_t pixel_count;
+    uint16_t pixel_array_length;
+    uint8_t *pixels;
+};
+
+static void
+led_strip_controler_close(struct sol_flow_node *node, void *data)
+{
+    struct lcd_strip_data *mdata = data;
+
+    if (mdata->spi)
+        sol_spi_close(mdata->spi);
+}
+
+static int
+led_strip_controler_open(struct sol_flow_node *node, void *data, const struct sol_flow_node_options *options)
+{
+    struct lcd_strip_data *mdata = data;
+    const struct sol_flow_node_type_adafruit_led_strip_options *opts;
+    uint16_t dataBytes;
+    uint8_t latchBytes;
+
+    opts = (const struct sol_flow_node_type_adafruit_led_strip_options *)options;
+
+    SOL_INT_CHECK(opts->pixel_count.val, < 0, -EINVAL);
+    mdata->pixel_count = opts->pixel_count.val;
+
+    dataBytes = mdata->pixel_count * 3;
+    latchBytes = (mdata->pixel_count + 31) / 32;
+    mdata->pixel_array_length = dataBytes + latchBytes;
```

Comment:
Since there's possibility of `data_bytes + latch_bytes` overflowing an `uint16_t`, it might be better to save this to an `uint32_t`, and checking if it's over `UINT16_MIN` before attributing this value to `pixel_array_length`.